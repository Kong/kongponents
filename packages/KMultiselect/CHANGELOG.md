# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [4.0.1](https://github.com/Kong/kongponents/compare/v4.0.0...v4.0.1) (2021-09-07)

**Note:** Version bump only for package @kongponents/kmultiselect





# 4.0.0 (2021-09-03)


### Features

* **KMultiselect:** Adds multichecklist component ‚úîÔ∏èüöÄ ([#217](https://github.com/Kong/kongponents/issues/217)) ([f8354e0](https://github.com/Kong/kongponents/commit/f8354e07163a81baba5f68a118bf52acd167542e))
* **style:** Updates Colors & Component Styles ([#225](https://github.com/Kong/kongponents/issues/225)) ([642a26b](https://github.com/Kong/kongponents/commit/642a26b888cbf8b4abbe32965729d44d54f9dfef))


### BREAKING CHANGES

* **style:** All new colors and color naming. Various component style improvements.





## [1.0.4](https://github.com/Kong/kongponents/compare/@kongponents/kmultiselect@1.0.3...@kongponents/kmultiselect@1.0.4) (2020-12-04)

**Note:** Version bump only for package @kongponents/kmultiselect





## [1.0.3](https://github.com/Kong/kongponents/compare/@kongponents/kmultiselect@1.0.2...@kongponents/kmultiselect@1.0.3) (2020-10-08)

**Note:** Version bump only for package @kongponents/kmultiselect





## [1.0.2](https://github.com/Kong/kongponents/compare/@kongponents/kmultiselect@1.0.1...@kongponents/kmultiselect@1.0.2) (2020-07-13)

**Note:** Version bump only for package @kongponents/kmultiselect





## [1.0.1](https://github.com/Kong/kongponents/compare/@kongponents/kmultiselect@1.0.0...@kongponents/kmultiselect@1.0.1) (2020-06-30)

**Note:** Version bump only for package @kongponents/kmultiselect





# [1.0.0](https://github.com/Kong/kongponents/compare/@kongponents/kmultiselect@0.3.0...@kongponents/kmultiselect@1.0.0) (2020-06-25)


### Features

* **style:** Updates Colors & Component Styles ([#225](https://github.com/Kong/kongponents/issues/225)) ([642a26b](https://github.com/Kong/kongponents/commit/642a26b888cbf8b4abbe32965729d44d54f9dfef))


### BREAKING CHANGES

* **style:** All new colors and color naming. Various component style improvements.





# 0.3.0 (2020-05-21)


### Features

* **KMultiselect:** Adds multichecklist component ‚úîÔ∏èüöÄ ([#217](https://github.com/Kong/kongponents/issues/217)) ([f8354e0](https://github.com/Kong/kongponents/commit/f8354e07163a81baba5f68a118bf52acd167542e))





# 0.2.0 (2020-05-21)


### Features

* **KMultiselect:** Adds multichecklist component ‚úîÔ∏èüöÄ ([#217](https://github.com/Kong/kongponents/issues/217)) ([f8354e0](https://github.com/Kong/kongponents/commit/f8354e07163a81baba5f68a118bf52acd167542e))
