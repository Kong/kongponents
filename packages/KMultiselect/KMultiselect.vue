<template>
  <KPop
    :width="width"
    :popover-timeout="0"
    :hide-popover="hidePopover"
    hide-caret
    placement="bottomStart"
    popover-classes="k-multiselect mt-0 mb-0"
    @opened="handleOpen"
    @closed="handleClose">
    <KButton
      v-bind="buttonAttributes"
      :appearance="buttonAttributes.appearance || 'secondary'"
      class="k-multiselect-trigger">
      {{ buttonText }}
    </KButton>
    <template
      v-if="!hidePopover"
      #content>
      <div class="k-multiselect-dropdown">
        <div
          v-if="title"
          class="k-multiselect-title">
          {{ title }}
        </div>
        <div
          v-if="hasFilter"
          class="k-multiselect-filter">
          <input
            ref="filterInput"
            v-model="filter"
            type="search"
            class="k-input w-100"
            placeholder="Filter..." >
        </div>
        <div class="k-multiselect-list">
          <ul
            class="k-multiselect-menu"
            role="listbox"
            aria-labelledby="dropdown">
            <li
              v-for="item in filteredItems"
              :key="item.label"
              :class="{ 'is-selected': item.selected, 'is-disabled': item.disabled }"
              :disabled="item.disabled"
              class="k-multiselect-item"
              role="option"
              @click="handleClick(item)">
              {{ item.label }}
            </li>
          </ul>
          <div class="k-multiselect-footer">
            <KButton
              :disabled="applyDisabled"
              size="small"
              appearance="secondary"
              class="k-multiselect-apply"
              @click="handleApply">{{ applyText }}</KButton>
          </div>
        </div>
      </div>
    </template>
  </KPop>
</template>

<script>
import KPop from '@kongponents/kpop/KPop'
import KButton from '@kongponents/kbutton/KButton.vue'

export default {
  name: 'KMultiselect',
  components: { KButton, KPop },
  props: {
    width: {
      type: String,
      default: 'auto'
    },
    buttonText: {
      type: String,
      default: ''
    },
    applyText: {
      type: String,
      default: 'Apply'
    },
    buttonAttributes: {
      type: Object,
      default: () => ({})
    },
    title: {
      type: String,
      default: ''
    },
    hasFilter: {
      type: Boolean,
      default: false
    },
    items: {
      type: Array,
      default: () => [],
      // Items must have a label and a selected property
      validator: (items) => items
        .some(i => i.hasOwnProperty('label') && i.hasOwnProperty('selected'))
    }
  },

  data () {
    return {
      filter: '',
      hidePopover: false,
      internalItems: this.copyItems(this.items)
    }
  },

  computed: {
    filteredItems () {
      const filter = this.filter

      return this.internalItems.filter(i => i.label.toLowerCase().includes(filter.toLowerCase()))
    },

    applyDisabled () {
      const internalItems = this.internalItems
      const items = this.items

      for (var i = 0; i < internalItems.length; ++i) {
        if (internalItems[i].selected !== items.find(item => item.label === internalItems[i].label).selected) {
          return false
        }
      }

      return true
    }
  },

  methods: {
    handleClick (item) {
      item.selected = !item.selected
    },

    handleApply () {
      this.hidePopover = true
      this.$emit('changes', this.copyItems(this.internalItems))
    },

    async handleOpen () {
      if (!this.hasFilter) return

      // Wait for popper to open & position itself
      await this.$nextTick()
      await this.$nextTick()
      this.$refs.filterInput.focus()
    },

    async handleClose () {
      await this.$nextTick()
      this.hidePopover = false
    },

    copyItems (items) {
      return items.map(obj => ({...obj}))
    }
  }
}
</script>

<style scoped lang="scss">
@import '~@kongponents/styles/_variables.scss';
@import '~@kongponents/styles/forms/_inputs.scss';

.k-multiselect {
  &-trigger {
    position: relative;
    display: inline-block;
    &:after {
      display: inline-block;
      width: 0;
      height: 0;
      margin-left: var(--spacing-xs, spacing(xs));
      vertical-align: middle;
      content: "";
      border-top: 0.325em solid;
      border-right: 0.325em solid transparent;
      border-left: 0.325em solid transparent;
    }
    &:disabled { pointer-events: none; }
  }
  &-filter .k-input {
    --KInputBackground: var(--grey-98, color(grey-98));
    --KInputBorder: var(--grey-88, color(grey-88));

    border-radius: 0;
    border-left: 0;
    border-right: 0;
    margin-bottom: var(--spacing-sm, spacing(sm));
    &:focus {
      border-color: var(--KInputBorder);
      border-width: 1px 0 1px 0;
    }
  }
  &-dropdown {
    padding: 1rem 0;
  }
  &-title {
    padding: 0 2rem;
    margin-bottom: var(--spacing-sm);
    color: var(--KMultiselectTitleColor, var(--black-45, color(black-45)));
    font-size: var(--KMultiselectTitleSize, var(--type-xs, type(xs)));
  }
  &-menu {
    padding: 0;
    margin: 0;
    list-style: none;
  }
  &-item {
    position: relative;
    padding: var(--KMultiselectItemPaddingY, 0) var(--KMultiselectItemPaddingX, var(--spacing-xl, spacing(xl)));
    white-space: nowrap;
    color: var(--KMultiselectItemColor, var(--black-70, color(black-70)));
    font-size: 1rem;
    line-height: 2rem;
    overflow: hidden;
    text-overflow: ellipsis;
    cursor: pointer;
    &:hover { background-color: var(--KMultiselectItemHover, var(--blue-100, color(blue-100)));}
    &.is-selected {
      color: var(--KMultiselectItemActiveColor, var(--blue-700, color(blue-700)));
      &:before {
        position: absolute;
        top: 10px;
        display: block;
        width: 12px;
        height: 12px;
        margin-left: -1.25rem;
        background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='16' height='13'%3E%3Cpath fill='%2366BFFF' fill-rule='evenodd' d='M4.5 10L13.7968194.7031806c.3883559-.3883559 1.021292-.3850692 1.4092576.0028964l.087846.087846c.3899555.3899555.3864986 1.0256554.0040118 1.4081422L5.2020652 12.2979348c-.3877399.3877399-1.0256941.3784363-1.4051902-.0010598l-3.09375-3.09375C.3147998 8.8147998.3181114 8.1818886.706077 7.793923l.087846-.087846c.3899555-.3899555 1.0271296-.3850244 1.4170351.0048811L4.5 10z'/%3E%3C/svg%3E");
        background-repeat: no-repeat;
        background-size: 100%;
        content: '';
      }
    }
    &.is-disabled {
      color: var(--KMultiselectDisabledColor, var(--black-45, color(black-45)));
      pointer-events: none;
    }
  }
  &-footer { margin: var(--spacing-sm, spacing(sm)) var(--spacing-xl, spacing(xl)) 0; }
}
</style>

<style>
.k-multiselect {
  --KPopPaddingY: 0;
  --KPopPaddingX: 0;
}
.k-dropdown .k-multiselect:focus,
.k-dropdown .k-multiselect:active {
  outline: 0;
}
</style>
