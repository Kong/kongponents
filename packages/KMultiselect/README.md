# @kongponents/kmultiselect

[![](https://img.shields.io/npm/v/@kongponents/kmultiselect.svg?style=flat-square)](https://www.npmjs.com/package/@kongponents/kmultiselect)

```html
<KMultiselect
  toggle-text="Items"
  :items="[{label: 'one', selected: false}, {label: 'two', selected: false}, {label: 'three', selected: true }]"/>
```
