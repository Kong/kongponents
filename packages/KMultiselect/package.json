{
  "name": "@kongponents/kmultiselect",
  "version": "1.0.0",
  "description": "Multiselect dropdown component",
  "main": "dist/KMultiselect.umd.min.js",
  "componentName": "KMultiselect",
  "source": "KMultiselect.vue",
  "files": [
    "dist",
    "KMultiselect.vue"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@kongponents/kbutton": "^1.0.0",
    "@kongponents/kpop": "^1.0.0"
  }
}
