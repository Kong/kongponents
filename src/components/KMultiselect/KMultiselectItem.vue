<template>
  <div
    :key="item.key"
    :data-testid="`k-multiselect-item-${item.value}`"
    class="k-multiselect-item"
    @click="handleClick"
  >
    <div
      role="option"
      class="d-block"
    >
      <button
        :class="{ disabled, selected: item.selected }"
        :value="item.value"
      >
        <span class="k-multiselect-item-label mr-2">
          <slot name="content">{{ item.label }}</slot>
        </span>
        <span class="k-multiselect-selected-icon-container">
          <KIcon
            v-if="item.selected"
            class="selected-item-icon"
            icon="check"
            color="var(--blue-200)"
          />
        </span>
      </button>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import KIcon from '@/components/KIcon/KIcon.vue'

export default defineComponent({
  name: 'MultiselectItem',
  components: { KIcon },
  props: {
    item: {
      type: Object,
      default: null,
      // Items must have a label and value
      validator: (item: Record<string, number | string>): boolean => item.label !== undefined && item.value !== undefined,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
  },
  emits: ['selected'],
  setup(props, { emit }) {
    const handleClick = (): void => {
      emit('selected', props.item)
    }

    return {
      handleClick,
    }
  },
})
</script>

<style lang="scss" scoped>
@import '@/styles/variables';
@import '@/styles/functions';

.k-multiselect-item {
  margin-bottom: 4px;
  list-style: none !important;

  &:last-of-type {
    margin-bottom: 0 !important;
  }

  button {
    display: flex;
    align-items: center;
    font-size: var(--type-xs);
    line-height: 26px;
    color: var(--grey-500);
    padding-left: var(--spacing-xxs);
    width: 100%;
    border: none;
    font-family: var(--font-family-sans);
    background-color: var(--white);
    border-radius: 4px;
    text-align: left;
    font-weight: 400;

    &:not(:disabled),
    &:not(.disabled) {
      cursor: pointer;
    }

    .k-multiselect-item-label {
      width: auto;
      line-height: 16px;
      color: var(--grey-600);
      font-weight: 500;
      font-size: 14px;
      padding: 8px;
      margin-bottom: 4px;

      :deep(.select-item-label) {
        color: var(--grey-600);
        font-weight: 600;
        font-size: 14px;
        margin-bottom: 4px;
      }

      :deep(.select-item-desc) {
        color: var(--grey-500);
        font-weight: 400;
        font-size: 14px;
      }
    }

    .kong-icon:not(.selected-item-icon) {
      margin-right: .75rem;
    }

    .k-multiselect-selected-icon-container {
      margin-left: auto;
      margin-top: auto;
      margin-bottom: auto;
      height: 24px;
      width: 24px;

      .kong-icon {
        position: relative;
        top: 0;
        right: 0;
        transform: none;
      }
    }

    &:hover {
      background-color: var(--grey-100);
      color: var(--grey-600);
    }

    &.selected {
      background-color: var(--blue-100);
      color: var(--blue-500);
      font-weight: 400;
    }

    &.danger {
      color: var(--red-500);
    }
  }
}
</style>
