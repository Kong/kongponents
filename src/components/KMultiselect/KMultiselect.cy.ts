import { mount } from 'cypress/vue'
import { h } from 'vue'
import KMultiselect from '@/components/KMultiselect/KMultiselect.vue'

/**
 * ALL TESTS MUST USE testMode: true
 * We generate unique IDs for reference by aria properties. Test mode strips these out
 * allowing for successful snapshot verification.
 * props: {
 *   testMode: true
 * }
 */

describe('KMultiselect', () => {
  it('renders props when passed', () => {
    const labels = ['Label 1', 'Label 2', 'Label 3']
    const vals = ['label1', 'label2', 'label3']

    mount(KMultiselect, {
      props: {
        testMode: true,
        items: [{
          label: labels[0],
          value: vals[0],
        }, {
          label: labels[1],
          value: vals[1],
        }, {
          label: labels[2],
          value: vals[2],
        }],
      },
    })

    cy.get('.k-multiselect-input').trigger('click')

    cy.getTestId(`k-multiselect-item-${vals[0]}`).should('contain.text', labels[0])
    cy.getTestId(`k-multiselect-item-${vals[1]}`).should('contain.text', labels[1])
    cy.getTestId(`k-multiselect-item-${vals[2]}`).should('contain.text', labels[2])
    cy.get('.k-multiselect-popover').should('be.visible')
    cy.get('.k-multiselect-dropdown-footer-text').should('not.exist')
  })

  it('renders with selected items when focused', () => {
    const selectedLabel = 'Label 1'
    const selectedLabel2 = 'Label 2'

    mount(KMultiselect, {
      props: {
        testMode: true,
        items: [
          { label: selectedLabel, value: 'label1', selected: true },
          { label: selectedLabel2, value: 'label2', selected: true },
        ],
      },
    })

    cy.get('.k-multiselect-input input').should('have.attr', 'placeholder', '2 items selected')

    cy.get('.k-multiselect-input').trigger('click')

    cy.getTestId('k-multiselect-selections').should('contain.text', selectedLabel)
    cy.getTestId('k-multiselect-selections').should('contain.text', selectedLabel2)
  })

  it('renders with disabled item', () => {
    const labels = ['Label 1', 'Label 2', 'Label 3']
    const vals = ['label1', 'label2', 'label3']

    mount(KMultiselect, {
      props: {
        testMode: true,
        items: [
          { label: labels[0], value: vals[0], disabled: true },
          { label: labels[1], value: vals[1] },
        ],
      },
    })

    cy.get('.k-multiselect-input').trigger('click')

    cy.get(`[data-testid="k-multiselect-item-${vals[0]}"] button`).should('have.attr', 'disabled')
  })

  it('renders with correct px width', () => {
    const width = 350

    mount(KMultiselect, {
      props: {
        testMode: true,
        width: width + '',
        items: [{
          label: 'Label 1',
          value: 'label1',
          selected: true,
        }],
      },
    })

    cy.get('.k-multiselect').invoke('outerWidth').should('eq', width)
  })

  it('renders with correct label', () => {
    const labelText = 'Cool Beans!'

    mount(KMultiselect, {
      props: {
        testMode: true,
        label: labelText,
        items: [{
          label: 'Label 1',
          value: 'label1',
        }],
      },
    })

    cy.get('.k-input-label').should('contain.text', labelText)
  })

  it('renders label with labelAttributes applied', () => {
    const labelText = 'A Label'
    mount(KMultiselect, {
      props: {
        testMode: true,
        label: labelText,
        labelAttributes: {
          help: 'some help text',
        },
        items: [{
          label: 'Label 1',
          value: 'label1',
        }],
      },
    })

    cy.get('.k-input-label').should('contain.text', labelText)
    cy.get('.k-input-label .kong-icon-help').should('be.visible')
  })

  it('reacts to text change and select', () => {
    const labels = ['Label 1', 'Label 2']
    const vals = ['label1', 'label2']

    mount(KMultiselect, {
      props: {
        testMode: true,
        items: [{
          label: labels[0],
          value: vals[0],
        }, {
          label: labels[1],
          value: vals[1],
        }],
      },
    })

    cy.get('.k-multiselect-input').click()

    cy.getTestId(`k-multiselect-item-${vals[0]}`).should('contain.text', labels[0])
    cy.getTestId(`k-multiselect-item-${vals[1]}`).should('contain.text', labels[1])

    cy.get('input').type(labels[0])

    cy.getTestId(`k-multiselect-item-${vals[0]}`).should('contain.text', labels[0])
    cy.getTestId(`k-multiselect-item-${vals[1]}`).should('not.exist')

    cy.getTestId(`k-multiselect-item-${vals[0]}`).eq(0).click()
    cy.getTestId('k-multiselect-selections').should('contain.text', labels[0])
  })

  it('allows adding an item with enableItemCreation', () => {
    const labels = ['Label 1', 'Label 2']
    const vals = ['label1', 'label2']
    const newItem = 'Rock me'

    mount(KMultiselect, {
      props: {
        testMode: true,
        items: [{
          label: labels[0],
          value: vals[0],
        }, {
          label: labels[1],
          value: vals[1],
        }],
        enableItemCreation: true,
      },
    })

    cy.get('.k-multiselect-input').click()

    cy.getTestId(`k-multiselect-item-${vals[0]}`).should('contain.text', labels[0])
    cy.getTestId(`k-multiselect-item-${vals[1]}`).should('contain.text', labels[1])
    // no adding a label that already exists
    cy.get('input').type(labels[0])
    cy.getTestId('k-multiselect-add-item').should('not.exist')
    cy.get('input').clear()
    // add new item
    cy.get('input').type(newItem)
    cy.getTestId('k-multiselect-add-item').should('contain.text', newItem).click()
    // search is cleared
    cy.get('input').should('not.contain.text', newItem)
    // item displays in selections
    cy.getTestId('k-multiselect-selections').should('contain.text', newItem)
    // item displays when searching
    cy.get('input').type(newItem)
    cy.get('.k-multiselect-item .k-multiselect-item-label').should('contain.text', newItem)
    // no adding a label that already exists
    cy.getTestId('k-multiselect-add-item').should('not.exist')
    // item gone when dismissed
    cy.getTestId('k-multiselect-selections').get('.k-badge-dismiss-button').first().click()
    // removed from selections
    cy.getTestId('k-multiselect-selections').should('not.to.exist')
    // gone when searching
    cy.get('input').clear()
    cy.get('input').type(newItem)
    cy.get('.k-multiselect-item .selected .k-multiselect-item-label').should('not.exist')
  })

  it('clears added items when clicking clear all with enableItemCreation', () => {
    const labels = ['Label 1', 'Label 2']
    const vals = ['label1', 'label2']
    const newItem = 'Rock me'

    mount(KMultiselect, {
      props: {
        testMode: true,
        items: [{
          label: labels[0],
          value: vals[0],
        }, {
          label: labels[1],
          value: vals[1],
        }],
        enableItemCreation: true,
      },
    })

    cy.get('.k-multiselect-input').click()

    cy.getTestId(`k-multiselect-item-${vals[0]}`).should('contain.text', labels[0])
    cy.getTestId(`k-multiselect-item-${vals[1]}`).should('contain.text', labels[1])

    // add new item
    cy.get('input').type(newItem)
    cy.getTestId('k-multiselect-add-item').should('contain.text', newItem).click()
    // item displays in selections
    cy.getTestId('k-multiselect-selections').should('contain.text', newItem)
    cy.getTestId('k-multiselect-clear-icon').click()
    // cleared
    cy.getTestId('k-multiselect-selections').should('not.to.exist')
    cy.get('input').type(newItem)
    cy.get('.k-multiselect-item .selected .k-multiselect-item-label').should('not.exist')
  })

  it('ignores clicks on disabled item', () => {
    const labels = ['Label 1', 'Label 2']
    const vals = ['label1', 'label2']

    mount(KMultiselect, {
      props: {
        testMode: true,
        items: [{
          label: labels[0],
          value: vals[0],
          disabled: true,
        }, {
          label: labels[1],
          value: vals[1],
        }],
      },
    })

    cy.get('.k-multiselect-input').click()

    cy.getTestId(`k-multiselect-item-${vals[0]}`).click()
    cy.getTestId('k-multiselect-selections').should('not.exist')
  })

  it('allows slotting content into the items', async () => {
    const itemSlotContent = 'I am slotted baby!'
    const itemLabel = 'Label 1'
    const itemValue = 'label1'

    mount(KMultiselect, {
      props: {
        testMode: true,
        items: [{
          label: itemLabel,
          value: itemValue,
        }],
      },
      slots: {
        'item-template': h('span', {}, itemSlotContent),
      },
    })

    cy.getTestId(`k-multiselect-item-${itemValue}`).should('contain.text', itemSlotContent)
  })

  it('works in autosuggest mode', () => {
    const onQueryChange = cy.spy().as('onQueryChange')
    mount(KMultiselect, {
      props: {
        testMode: true,
        autosuggest: true,
        loading: false,
        items: [],
        onQueryChange,
      },
    })

    cy.get('input').type('a').then(() => {
      cy.get('@onQueryChange').should('have.been.calledWith', 'a')
    }).then(() => {
      cy.wrap(Cypress.vueWrapper.setProps({ loading: true })).get('.k-multiselect-chevron-icon.kong-icon-spinner').should('exist')
    }).then(() => {
      cy.wrap(Cypress.vueWrapper.setProps({ loading: false })).get('.k-multiselect-chevron-icon.kong-icon-spinner').should('not.exist')
    }).then(() => {
      cy.wrap(Cypress.vueWrapper.setProps({ items: [{ label: 'Label 1', value: 'label1' }] })).getTestId('k-multiselect-item-label1').should('contain.text', 'Label 1')
    })
  })

  it('counts invisible selected items correctly', () => {
    const allItems = Array.from(new Array(100)).map((_, i) => ({
      label: `Item ${i}`,
      value: `${i}`,
    }))

    const onQueryChange = cy.spy().as('onQueryChange')

    const selected = (Array.from(new Array(10)).map((_, i) => `${i}`))

    const items = (allItems.slice(0, 10))

    mount(KMultiselect, {
      props: {
        testMode: true,
        autosuggest: true,
        collapsedContext: true,
        selectedRowCount: 2,
        modelValue: selected,
        loading: false,
        items,
        onQueryChange,
      },
    })

    cy.get('[data-testid="k-multiselect-trigger"]')
      .click('topRight')
      .then(() => {
        cy.get('[data-testid="k-multiselect-selections"] > div .k-badge-text')
          .last()
          .should(($el) => {
            const text = $el.text()

            expect(text.trim()).to.equal('+7')
          })
      })
      .then(() => {
        cy.get('input').focus()
      })
      .then(() => {
        cy.get('@onQueryChange').should('have.been.calledWith', '')
        Cypress.vueWrapper.setProps({
          items: allItems.slice(5, 20),
        })
      })
      .then(() => {
        cy.get('input').type('{esc}')
      })
      .then(() => {
        cy.get('[data-testid="k-multiselect-selections"] > div .k-badge-text')
          .last()
          .should(($el) => {
            const text = $el.text()

            expect(text.trim()).to.equal('+7')
          })
      })
  })

  it('always shows selections when expandSelected is true', () => {
    const labels = ['Label 1', 'Label 2']
    const vals = ['label1', 'label2']

    mount(KMultiselect, {
      props: {
        testMode: true,
        expandSelected: true,
        items: [{
          label: labels[0],
          value: vals[0],
          selected: true,
        }, {
          label: labels[1],
          value: vals[1],
          selected: true,
        }],
      },
    })

    cy.getTestId('k-multiselect-selections').should('exist')

    cy.getTestId('k-multiselect-selections').should('contain.text', labels[0])
    cy.getTestId('k-multiselect-selections').should('contain.text', labels[1])

    cy.get('.k-multiselect-chevron-icon').click()

    cy.getTestId('k-multiselect-selections').should('exist')
  })

  it('always shows selections when collapsedContext is true', () => {
    const labels = ['Label 1', 'Label 2']
    const vals = ['label1', 'label2']

    mount(KMultiselect, {
      props: {
        testMode: true,
        collapsedContext: true,
        items: [{
          label: labels[0],
          value: vals[0],
          selected: true,
        }, {
          label: labels[1],
          value: vals[1],
          selected: true,
        }],
      },
    })

    cy.getTestId('k-multiselect-selections').should('exist')

    cy.getTestId('k-multiselect-selections').should('contain.text', labels[0])
    cy.getTestId('k-multiselect-selections').should('contain.text', labels[1])

    cy.get('.k-multiselect-chevron-icon').click()

    cy.getTestId('k-multiselect-selections').should('exist')
  })

  it('can clear all selections when focused', () => {
    const labels = ['Label 1', 'Label 2']
    const vals = ['label1', 'label2']

    mount(KMultiselect, {
      props: {
        testMode: true,
        items: [{
          label: labels[0],
          value: vals[0],
          selected: true,
        }, {
          label: labels[1],
          value: vals[1],
          selected: true,
        }],
      },
    })

    cy.get('.k-multiselect-input').click()

    cy.getTestId('k-multiselect-selections').should('contain.text', labels[0])
    cy.getTestId('k-multiselect-selections').should('contain.text', labels[1])
    cy.get('.k-multiselect-clear-icon').trigger('click')
    cy.getTestId('k-multiselect-selections').should('not.exist')
  })

  it('can clear selection by badge dismiss when focused', () => {
    const labels = ['Label 1', 'Label 2']
    const vals = ['label1', 'label2']

    mount(KMultiselect, {
      props: {
        testMode: true,
        items: [{
          label: labels[0],
          value: vals[0],
          selected: true,
        }, {
          label: labels[1],
          value: vals[1],
        }],
      },
    })

    cy.get('.k-multiselect-input').click()

    cy.getTestId('k-multiselect-selections').should('contain.text', labels[0])
    cy.getTestId('k-multiselect-selections').get('.k-badge-dismiss-button').first().click()
    cy.getTestId('k-multiselect-selections').should('not.exist')
  })

  it('renders dropdown footer text when prop is passed', () => {
    const labels = ['Label 1', 'Label 2', 'Label 3']
    const vals = ['label1', 'label2', 'label3']
    const dropdownFooterText = 'Dropdown footer text'

    mount(KMultiselect, {
      props: {
        testMode: true,
        items: [{
          label: labels[0],
          value: vals[0],
        }, {
          label: labels[1],
          value: vals[1],
        }, {
          label: labels[2],
          value: vals[2],
        }],
        dropdownFooterText,
      },
    })

    cy.get('.k-multiselect-input').trigger('click')

    cy.get('.k-multiselect-dropdown-footer-text').should('be.visible').should('contain.text', dropdownFooterText)
  })

  it('should allow slotting dropdown footer text', () => {
    const labels = ['Label 1', 'Label 2', 'Label 3']
    const vals = ['label1', 'label2', 'label3']
    const dropdownFooterText = 'Dropdown footer text'

    mount(KMultiselect, {
      props: {
        testMode: true,
        items: [{
          label: labels[0],
          value: vals[0],
        }, {
          label: labels[1],
          value: vals[1],
        }, {
          label: labels[2],
          value: vals[2],
        }],
        dropdownFooterText: 'This is getting replaced',
      },
      slots: {
        'dropdown-footer-text': dropdownFooterText,
      },
    })

    cy.get('.k-multiselect-input').trigger('click')

    cy.get('.k-multiselect-dropdown-footer-text').should('be.visible').should('contain.text', dropdownFooterText)
  })

  it('renders group titles and groups items in correct order', () => {
    const group1Title = 'Group 1'
    const group2Title = 'Group 2'
    const items = [
      { label: 'Label 0', value: 'value0' },
      { label: 'Label 1', value: 'value1', group: group1Title },
      { label: 'Label 3', value: 'value3', group: group2Title },
      { label: 'Label 2', value: 'value2', group: group1Title },
      { label: 'Label 4', value: 'value4', group: group2Title },
    ]

    mount(KMultiselect, {
      props: {
        testMode: true,
        items,
      },
    })

    cy.getTestId('k-multiselect-input').trigger('click')
    cy.get('.k-multiselect-item').eq(0).should('contain.text', items[0].label)
    cy.get('.k-multiselect-group-title').eq(0).should('contain.text', group1Title)
    cy.get('.k-multiselect-group-title').eq(1).should('contain.text', group2Title)
    cy.get('.k-multiselect-item').eq(1).should('contain.text', items[1].label)
    cy.get('.k-multiselect-item').eq(2).should('contain.text', items[3].label)
    cy.get('.k-multiselect-item').eq(3).should('contain.text', items[2].label)
    cy.get('.k-multiselect-item').eq(4).should('contain.text', items[4].label)
  })

  it('should able to handle tons of items with no obvious lag', () => {
    const items = Array.from(new Array(500)).map((_, i) => ({
      label: `Item ${i}`,
      value: `${i}`,
      selected: i < 400,
    }))

    const startTime = Date.now()

    mount(KMultiselect, {
      props: {
        testMode: true,
        items,
      },
    }).then(() => {
      expect(Date.now() - startTime).to.be.lessThan(3000)
    })
  })
})
