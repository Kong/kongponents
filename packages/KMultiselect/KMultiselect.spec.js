import { mount } from '@vue/test-utils'
import { default as KMultiselect } from './KMultiselect.vue'

const DEFAULT_PROPS = {
  items: [
    { label: 'Item 1', selected: true },
    { label: 'Item 2', selected: false }
  ]
}

describe('KMultiselect', () => {
  it('correctly adds active class', () => {
    const wrapper = mount(KMultiselect, {
      propsData: DEFAULT_PROPS
    })

    expect(wrapper.find('.k-multiselect-menu > li:first-child').classes('is-selected')).toBe(true)
  })

  it('emits proper updates on apply', async () => {
    const wrapper = mount(KMultiselect, {
      propsData: DEFAULT_PROPS
    })

    wrapper.find('.k-multiselect-trigger').trigger('click')
    wrapper.find('.k-multiselect-menu > li:last-child').trigger('click')
    wrapper.find('.k-multiselect-apply').trigger('click')

    await wrapper.vm.$nextTick()

    expect(wrapper.emitted().changes).toBeTruthy()
    expect(wrapper.emitted().changes[0][0][1]).toEqual({ label: 'Item 2', selected: true })
  })

  it('matches snapshot', () => {
    const wrapper = mount(KMultiselect, {
      propsData: DEFAULT_PROPS
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
