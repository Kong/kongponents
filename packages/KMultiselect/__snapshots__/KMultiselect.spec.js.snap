// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KMultiselect matches snapshot 1`] = `
<div id="test-target-id-1234" aria-controls="test-popover-id-1234" role="button"><button type="button" class="k-button k-multiselect-trigger has-caret medium rounded secondary has-caret">
    <!---->

    <span class="kong-icon kong-icon-chevronDown caret has-caret"><svg viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
  <title>Expand</title>
  <path d="m4.6 7 5.2 5L15 7" fill="none" stroke="#A3BBCC" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"></path>
</svg>
</span></button>
  <div id="test-popover-id-1234" role="region" class="k-popover k-multiselect mt-0 mb-0 hide-caret" style="width: auto; max-width: 350px; max-height: auto; display: none;" name="fade">
    <!---->
    <div class="k-popover-content">
      <div class="k-multiselect-dropdown">
        <!---->
        <!---->
        <div class="k-multiselect-list">
          <ul role="listbox" aria-labelledby="dropdown" class="k-multiselect-menu">
            <li role="option" class="k-multiselect-item is-selected">
              Item 1
            </li>
            <li role="option" class="k-multiselect-item">
              Item 2
            </li>
          </ul>
          <div class="k-multiselect-footer"><button type="button" class="k-button k-multiselect-apply small rounded secondary" disabled="disabled">
              <!----> Apply
              <!----></button></div>
        </div>
      </div>
    </div>
    <!---->
  </div>
</div>
`;
